File: document/src/vespa/document/datatype/mapdatatype.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "mapdatatype.h"
#include <vespa/document/fieldvalue/mapfieldvalue.h>
#include <vespa/vespalib/stllike/asciistream.h>
#include <ostream>

namespace document {

using namespace std::literals;

namespace {

constexpr auto key_keyword = "key"sv;
constexpr auto value_keyword = "value"sv;

vespalib::string createName(const DataType& keyType, const DataType& valueType)
{
    vespalib::asciistream ost;
    ost << "Map<" << keyType.getName() << "," << valueType.getName() << ">";
    return ost.str();
}
}  // namespace

MapDataType::MapDataType(const DataType &key, const DataType &value) noexcept
    : DataType(createName(key, value)),
      _keyType(&key),
      _valueType(&value) {
}

MapDataType::MapDataType(const DataType &key, const DataType &value, int id) noexcept
    : DataType(createName(key, value), id),
      _keyType(&key),
      _valueType(&value) {
}

MapDataType::~MapDataType() = default;

FieldValue::UP MapDataType::createFieldValue() const {
    return std::make_unique<MapFieldValue>(*this);
}

void
MapDataType::print(std::ostream& out, bool verbose,
                   const std::string& indent) const
{
    out << "MapDataType(";
    getKeyType().print(out, verbose, indent + "    ");
    out << ", ";
    getValueType().print(out, verbose, indent + "    ");
    out << ", id " << getId() << ")";
}

bool
MapDataType::equals(const DataType& other) const noexcept
{
    if (this == &other) return true;
    if (!DataType::equals(other)) return false;
    const MapDataType * w = other.cast_map();
    return w && _keyType->equals(*w->_keyType) && _valueType->equals(*w->_valueType);
}

void
MapDataType::buildFieldPathImpl(FieldPath & path, const DataType &dataType,
                                std::string_view remainFieldName,
                                const DataType &keyType, const DataType &valueType)
{
    if (!remainFieldName.empty() && remainFieldName[0] == '{') {
        std::string_view rest = remainFieldName;
        vespalib::string keyValue = FieldPathEntry::parseKey(rest);

        valueType.buildFieldPath(path, (!rest.empty() && rest[0] == '.') ? rest.substr(1) : rest);

        if (remainFieldName.size() > 1 && remainFieldName[1] == '$') {
            path.insert(path.begin(), std::make_unique<FieldPathEntry>(valueType, keyValue.substr(1)));
        } else {
            FieldValue::UP fv = keyType.createFieldValue();
            *fv = keyValue;
            path.insert(path.begin(), std::make_unique<FieldPathEntry>(valueType, dataType, std::move(fv)));
        }
    } else if (remainFieldName.starts_with(key_keyword)) {
        size_t endPos = key_keyword.size();;
        if (remainFieldName.size() > endPos && remainFieldName[endPos] == '.') {
            endPos++;
        }

        keyType.buildFieldPath(path, remainFieldName.substr(endPos));

        path.insert(path.begin(), std::make_unique<FieldPathEntry>(dataType, keyType, valueType, true, false));
    } else if (remainFieldName.starts_with(value_keyword)) {
        size_t endPos = value_keyword.size();
        if (remainFieldName.size() > endPos && remainFieldName[endPos] == '.') {
            endPos++;
        }

        valueType.buildFieldPath(path, remainFieldName.substr(endPos));

        path.insert(path.begin(), std::make_unique<FieldPathEntry>(dataType, keyType, valueType, false, true));
    } else {
        keyType.buildFieldPath(path, remainFieldName);
    }
}

void
MapDataType::onBuildFieldPath(FieldPath & fieldPath, std::string_view remainFieldName) const
{
    buildFieldPathImpl(fieldPath, *this, remainFieldName, getKeyType(), getValueType());
}

}  // namespace document

File: document/src/main/java/com/yahoo/document/MapDataType.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.document;

import com.yahoo.document.datatypes.FieldValue;
import com.yahoo.document.datatypes.MapFieldValue;

import java.util.ArrayList;
import java.util.List;

/**
 * Represents a map type.
 *
 * @author vegardh
 */
public class MapDataType extends DataType {

    private DataType keyType;
    private DataType valueType;

    public MapDataType(DataType keyType, DataType valueType, int id) {
        super("Map<"+keyType.getName()+","+valueType.getName()+">", id);
        this.keyType=keyType;
        this.valueType = valueType;
    }

    public MapDataType(DataType keyType, DataType valueType) {
        this(keyType, valueType, 0);
        setId(getName().toLowerCase().hashCode());
    }

    @Override
    public MapDataType clone() {
        MapDataType type = (MapDataType)super.clone();
        type.keyType = keyType.clone();
        type.valueType = valueType.clone();
        return type;
    }

    @Override
    protected FieldValue createByReflection(Object arg) { return null; }

    @Override
    public boolean isValueCompatible(FieldValue value) {
        return value.getDataType().equals(this);
    }

    public DataType getKeyType() {
        return keyType;
    }

    public DataType getValueType() {
        return valueType;
    }

    @Override
    public MapFieldValue createFieldValue() {
        return new MapFieldValue(this);
    }

    @Override
    public Class getValueClass() {
        return MapFieldValue.class;
    }

    @Override
    protected void register(DocumentTypeManager manager,
            List<DataType> seenTypes) {
        seenTypes.add(this);
        if (!seenTypes.contains(getKeyType())) {
            getKeyType().register(manager, seenTypes);
        }
        if (!seenTypes.contains(getValueType())) {
            getValueType().register(manager, seenTypes);
        }
        super.register(manager, seenTypes);
    }

    public static FieldPath buildFieldPath(String remainFieldName, DataType keyType, DataType valueType) {
        if (remainFieldName.length() > 0 && remainFieldName.charAt(0) == '{') {
            FieldPathEntry.KeyParseResult result = FieldPathEntry.parseKey(remainFieldName);
            String keyValue = result.parsed;

            FieldPath path = valueType.buildFieldPath(skipDotInString(remainFieldName, result.consumedChars - 1));
            List<FieldPathEntry> tmpPath = new ArrayList<FieldPathEntry>(path.getList());

            if (remainFieldName.charAt(1) == '$') {
                tmpPath.add(0, FieldPathEntry.newVariableLookupEntry(keyValue.substring(1), valueType));
            } else {
                FieldValue fv = keyType.createFieldValue();
                fv.assign(keyValue);
                tmpPath.add(0, FieldPathEntry.newMapLookupEntry(fv, valueType));
            }

            return new FieldPath(tmpPath);

        } else if (remainFieldName.startsWith("key")) {
            FieldPath path = keyType.buildFieldPath(skipDotInString(remainFieldName, 2));
            List<FieldPathEntry> tmpPath = new ArrayList<FieldPathEntry>(path.getList());
            tmpPath.add(0, FieldPathEntry.newAllKeysLookupEntry(keyType));
            return new FieldPath(tmpPath);
        } else if (remainFieldName.startsWith("value")) {
            FieldPath path = valueType.buildFieldPath(skipDotInString(remainFieldName, 4));
            List<FieldPathEntry> tmpPath = new ArrayList<FieldPathEntry>(path.getList());
            tmpPath.add(0, FieldPathEntry.newAllValuesLookupEntry(valueType));
            return new FieldPath(tmpPath);
        }

        return keyType.buildFieldPath(remainFieldName);
    }

    @Override
    public FieldPath buildFieldPath(String remainFieldName) {
        return buildFieldPath(remainFieldName, getKeyType(), getValueType());
    }

    @Override
    public boolean isMultivalue() { return true; }

}
